File: benchmarks/src/perl/spectral_norm.pl
use strict;
use warnings;
use Time::HiRes qw(time);

sub eval_A {
    my ($i, $j) = @_;
    return 1.0 / ((($i + $j) * ($i + $j + 1) / 2) + $i + 1);
}

sub eval_A_times_u {
    my ($u, $v, $n) = @_;
    for my $i (0 .. $n - 1) {
        $v->[$i] = 0.0;
        for my $j (0 .. $n - 1) {
            $v->[$i] += eval_A($i, $j) * $u->[$j];
        }
    }
}

sub eval_At_times_u {
    my ($u, $v, $n) = @_;
    for my $i (0 .. $n - 1) {
        $v->[$i] = 0.0;
        for my $j (0 .. $n - 1) {
            $v->[$i] += eval_A($j, $i) * $u->[$j];
        }
    }
}

sub eval_AtA_times_u {
    my ($u, $v, $w, $n) = @_;
    eval_A_times_u($u, $w, $n);
    eval_At_times_u($w, $v, $n);
}

my $start = time();
my $n = 150;
my @u = (1.0) x $n;
my @v = (0.0) x $n;
my @w = (0.0) x $n;

for my $i (1 .. 10) {
    eval_AtA_times_u(\@u, \@v, \@w, $n);
    eval_AtA_times_u(\@v, \@u, \@w, $n);
}

my $vBv = 0.0;
my $vv = 0.0;
for my $i (0 .. $n - 1) {
    $vBv += $u[$i] * $v[$i];
    $vv += $v[$i] * $v[$i];
}

my $res = sqrt($vBv / $vv);
my $end = time();

printf("Result: %.9f\n", $res);
printf("Time: %dms\n", ($end - $start) * 1000);

File: benchmarks/src/php/spectral_norm.php
<?php
function eval_A($i, $j) {
    return 1.0 / ((($i + $j) * ($i + $j + 1) / 2) + $i + 1);
}

function eval_A_times_u($u, &$v, $n) {
    for ($i = 0; $i < $n; $i++) {
        $v[$i] = 0.0;
        for ($j = 0; $j < $n; $j++) {
            $v[$i] += eval_A($i, $j) * $u[$j];
        }
    }
}

function eval_At_times_u($u, &$v, $n) {
    for ($i = 0; $i < $n; $i++) {
        $v[$i] = 0.0;
        for ($j = 0; $j < $n; $j++) {
            $v[$i] += eval_A($j, $i) * $u[$j];
        }
    }
}

function eval_AtA_times_u($u, &$v, &$w, $n) {
    eval_A_times_u($u, $w, $n);
    eval_At_times_u($w, $v, $n);
}

$start = microtime(true);
$n = 150;
$u = array_fill(0, $n, 1.0);
$v = array_fill(0, $n, 0.0);
$w = array_fill(0, $n, 0.0);

for ($i = 0; $i < 10; $i++) {
    eval_AtA_times_u($u, $v, $w, $n);
    eval_AtA_times_u($v, $u, $w, $n);
}

$vBv = 0.0;
$vv = 0.0;
for ($i = 0; $i < $n; $i++) {
    $vBv += $u[$i] * $v[$i];
    $vv += $v[$i] * $v[$i];
}

$res = sqrt($vBv / $vv);
$end = microtime(true);

printf("Result: %.9f\n", $res);
printf("Time: %dms\n", ($end - $start) * 1000);
